R/preprocessing.R

```diff
@@ -177,3 +177,154 @@ default_fill_missing_obs <- function(data, obs, obs_column) {
   }
   return(data)
 }
+
+##' Add missing values for future dates
+##'
+##' @param accumulate The number of days to accumulate when generating posterior
+##'   prediction, e.g. 7 for weekly accumulated forecasts.
+##' @inheritParams add_horizon
+##' @inheritParams estimate_infections
+##' @importFrom data.table copy merge.data.table setDT
+##' @return A data.table with missing values for future dates
+##' @keywords internal
+add_horizon <- function(data, horizon, accumulate = 1L,
+                        obs_column = "confirm", by = NULL) {
+  assert_data_frame(data)
+  assert_character(obs_column)
+  assert_character(by, null.ok = TRUE)
+  assert_names(
+    colnames(data),
+    must.include = c("date", by, obs_column)
+  )
+  assert_integerish(horizon, lower = 0)
+  assert_integerish(accumulate, lower = 1)
+  assert_date(data$date, any.missing = FALSE)
+
+  reported_cases <- data.table::setDT(data)
+  if (horizon > 0) {
+    reported_cases_grid <- data.table::copy(reported_cases)[,
+      .(date = seq(max(date) + 1, max(date) + horizon, by = "days")),
+      by = by
+    ]
+    ## if we accumulate add the column
+    if (accumulate > 1 || "accumulate" %in% colnames(data)) {
+      reported_cases_grid[, accumulate := TRUE]
+      ## set accumulation to FALSE where appropriate
+      if (horizon >= accumulate) {
+        reported_cases_grid[
+          as.integer(date - min(date) - 1) %% accumulate == 0,
+          accumulate := FALSE
+        ]
+      }
+    }
+    ## fill any missing columns
+    reported_cases_grid <- data.table::merge.data.table(
+      reported_cases, reported_cases_grid,
+      by = "date", all.y = TRUE
+    )
+  }
+  return(reported_cases[])
+}
+
+##' Add breakpoints to certain dates in a data set.
+##'
+##' @param dates A vector of dates to use as breakpoints.
+##' @inheritParams estimate_infections
+##' @return A data.table with `breakpoint` set to 1 on each of the specified
+##'   dates.
+##' @export
+##' @importFrom data.table setDT
+##' @examples
+##' reported_cases <- add_breakpoints(example_confirmed, as.Date("2020-03-26"))
+add_breakpoints <- function(data, dates = as.Date(character(0))) {
+  assert_data_frame(data)
+  assert_names(colnames(data), must.include = "date")
+  assert_date(dates)
+  assert_date(data$date, any.missing = FALSE)
+  reported_cases <- data.table::setDT(data)
+  if (is.null(reported_cases$breakpoint)) {
+    reported_cases$breakpoint <- 0
+  }
+  missing_dates <- setdiff(dates, data$date)
+  if (length(missing_dates) > 0) {
+    cli_abort("Breakpoint date{?s} not found in data: {.var {missing_dates}}")
+  }
+  reported_cases[date %in% dates, breakpoint := 1]
+  return(reported_cases)
+}
+
+##' Filter leading zeros from a data set.
+##'
+##' @inheritParams estimate_infections
+##' @inheritParams fill_missing
+##' @return A data.table with leading zeros removed.
+##' @export
+##' @importFrom data.table setDT
+##' @examples
+##' cases <- data.table(
```

Comment:
```suggestion
##' cases <- data.frame(
```

as `data.table` isn't loaded this currently errors